### DISCUSSION TITLE
Segment duration calculation and IsError logic

### DISCUSSION THREAD

**Discussion by inversionhourglass** [apache/skywalking#10371] [Q&A]
I learned from [SegmentAnalysisListener](https://github.com/apache/skywalking/blob/master/oap-server/analyzer/agent-analyzer/src/main/java/org/apache/skywalking/oap/server/analyzer/provider/trace/parser/listener/SegmentAnalysisListener.java), that the total duration of a segment is the difference between the minimum `startTime` of all spans and the maximum `endTime` of all spans. Why not directly take the start time and end time of `firstSpan`. I'm writing a client agent, some AsyncSpan may start in main thread and finish in background thread, this AsyncSpan's `endTime` will be greater than `firstSpan`. In this case, the AsyncSpan's `endTime` will be apply to segment's `endTime` but I hope it use `firstSpan's endTime`. I know that I can create a new segment in new thread, but It's complicated.

Another question is about `Segment.isError`. It also learnd from [SegmentAnalysisListener](https://github.com/apache/skywalking/blob/master/oap-server/analyzer/agent-analyzer/src/main/java/org/apache/skywalking/oap/server/analyzer/provider/trace/parser/listener/SegmentAnalysisListener.java), if any span of segment is error then the segment will be error too. Every time a `service_sla` alarm is received, I will go to Service page and search traces with Error status condition. Search results will contains segments which only internal span failed. Can we add a tag for the segment if the endpoint failed, and then make this tag searchable?

**Comment by inversionhourglass** [comment]
OK, About the second question, the segment status is determined by [SegmentStatusStrategy](https://github.com/apache/skywalking/blob/master/oap-server/analyzer/agent-analyzer/src/main/java/org/apache/skywalking/oap/server/analyzer/provider/trace/parser/listener/strategy/SegmentStatusStrategy.java), the default strategy is `FROM_SPAN_STATUS`, it can be change by configuration, `SW_SEGMENT_STATUS_ANALYSIS_STRATEGY` of environment variables or `agent-analyzer:default:segmentStatusAnalysisStrategy` of the application.yml. But in my opinion, it is better to store the result of any span error as a tag, so that the failed segment can be filtered by status (status is determined by the strategy), and the segment containing the failed span can be filtered by tag.

About the first question, there is no answer so far, and the author is not ready to answer it. Maybe the question is too stupid, okay.

**Comment by wu-sheng** [comment]
SkyWalking as a OSS community, we don't  explain codes or guide people to change codes in private codebase. Generally because we can't afford the time.
That is why you may not see maintainers or me answer such detailed things.

**Reply by inversionhourglass** [reply]
okay, without discussing the codes. Can skywalking provide two filter conditions to filter out the failed segments and any span failed segments? Now you can only choose one of the two through configuration.
When the `service_sla` alarm occurs, it is generally hoped to filter out the segments of the alarm through the status condition (FROM_ENTRY_SPAN). It is also hoped that there is a way to view the segments of any span failed.

**Reply by wu-sheng** [reply]
I think you need to create two metrics based on different conditions and choose for different cases. 
Learn OAL, https://skywalking.apache.org/docs/main/next/en/concepts-and-designs/oal/, to create metrics you need.

**Reply by inversionhourglass** [reply]
This should have nothing to do with the metric. I want to filter out the segments of the entry span failed and filter out the segments of any span failed on the trace page. Now only one of them can be filtered out by status condition.

![image](https://user-images.githubusercontent.com/15099549/218795215-44ec3f6e-b8d1-470d-9b03-5f8b30f50ba0.png)

**Reply by wu-sheng** [reply]
You were saying about alerting. But show me the trace query page. I don't know what you mean.
From a UI perspective, you can see what you can do. I don't think I have more to share.

**Reply by inversionhourglass** [reply]
ok, thanks for your reply.
